drivers/spi/spi_ll_stm32.c

```diff
@@ -0,0 +1,457 @@
+/*
+ * Copyright (c) 2016 BayLibre, SAS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_SPI_LEVEL
+#include <logging/sys_log.h>
+
+#include <misc/util.h>
+#include <kernel.h>
+#include <board.h>
+#include <errno.h>
+#include <spi.h>
+
+#include <clock_control/stm32_clock_control.h>
+#include <clock_control.h>
+
+#include <drivers/spi/spi_ll_stm32.h>
+#include <spi_ll_stm32.h>
+
+#define CONFIG_CFG(cfg)						\
+((const struct spi_stm32_config * const)(cfg)->dev->config->config_info)
+
+#define CONFIG_DATA(cfg)					\
+((struct spi_stm32_data * const)(cfg)->dev->driver_data)
+
+#ifdef CONFIG_SPI_STM32_INTERRUPT
+static void spi_stm32_transmit(SPI_TypeDef *spi, struct spi_stm32_data *data);
+static void spi_stm32_receive(SPI_TypeDef *spi, struct spi_stm32_data *data);
+
+static void spi_stm32_isr(void *arg)
+{
+	struct device * const dev = (struct device *) arg;
+	const struct spi_stm32_config *cfg = dev->config->config_info;
+	struct spi_stm32_data *data = dev->driver_data;
+	SPI_TypeDef *spi = cfg->spi;
+
+	if (LL_SPI_IsActiveFlag_TXE(spi) && spi_context_tx_on(&data->ctx)) {
+		spi_stm32_transmit(spi, data);
+	}
+
+	if (LL_SPI_IsActiveFlag_RXNE(spi)) {
+		if (spi_context_rx_on(&data->ctx)) {
+			spi_stm32_receive(spi, data);
+		} else {
+			LL_SPI_DisableIT_RXNE(spi);
+		}
+	}
+
+	if (!spi_context_tx_on(&data->ctx) &&
+	    !spi_context_rx_on(&data->ctx)) {
+		LL_SPI_DisableIT_TXE(spi);
+		LL_SPI_DisableIT_RXNE(spi);
+
+		spi_context_cs_control(&data->ctx, false);
+
+#if defined(CONFIG_SOC_SERIES_STM32L4X) || defined(CONFIG_SOC_SERIES_STM32F3X)
+		/* Flush RX buffer */
+		while (LL_SPI_IsActiveFlag_RXNE(spi)) {
+			(void) LL_SPI_ReceiveData8(spi);
+		}
+#endif
+
+		if (LL_SPI_GetMode(spi) == LL_SPI_MODE_MASTER) {
+			while (LL_SPI_IsActiveFlag_BSY(spi)) {
+				/* NOP */
+			}
+			LL_SPI_Disable(spi);
+		}
+
+		spi_context_complete(&data->ctx, 0);
+	}
+}
+#endif
+
+static int spi_stm32_configure(struct spi_config *config)
+{
+	const struct spi_stm32_config *cfg = CONFIG_CFG(config);
+	struct spi_stm32_data *data = CONFIG_DATA(config);
+	const u32_t scaler[] = {
+		LL_SPI_BAUDRATEPRESCALER_DIV2,
+		LL_SPI_BAUDRATEPRESCALER_DIV4,
+		LL_SPI_BAUDRATEPRESCALER_DIV8,
+		LL_SPI_BAUDRATEPRESCALER_DIV16,
+		LL_SPI_BAUDRATEPRESCALER_DIV32,
+		LL_SPI_BAUDRATEPRESCALER_DIV64,
+		LL_SPI_BAUDRATEPRESCALER_DIV128,
+		LL_SPI_BAUDRATEPRESCALER_DIV256
+	};
+	SPI_TypeDef *spi = cfg->spi;
+	u32_t clock;
+	int br;
+
+	if (spi_context_configured(&data->ctx, config)) {
+		/* Nothing to do */
+		return 0;
+	}
+
+	if (SPI_WORD_SIZE_GET(config->operation) != 8) {
+		return -ENOTSUP;
+	}
+
+	clock_control_get_rate(device_get_binding(STM32_CLOCK_CONTROL_NAME),
+			       (clock_control_subsys_t) &cfg->pclken, &clock);
+
+	for (br = 1 ; br <= ARRAY_SIZE(scaler) ; ++br) {
+		u32_t clk = clock >> br;
+
+		if (clk < config->frequency) {
```

Comment:
Return -EINVAL if the given frequency is not supported, with a relevant logging message (go for SYS_LOG_ERR() in this case). Up to the user to know a little the hw specs he is working against.

Or, if the driver sets the highest speed it could, lower than requested, go for SYS_LOG_WARN() and don't return -EINVAL.

In any case, user should know what the hardware is supposed to support.